Module to update site lists for WMAgents

vkuznet opened the pull request:
Fixes #12039 

#### Status
In development

#### Description
Provide site list updates for WMAgents with the following logic:
        - obtain list of current active workflows from the agent using WMBS database
           - we extract workflow name and pickle file name (the spec)
        - requests their JSON specs from upstream wmstats server
        - check and if necessary update site lists of all active workflows
           - push new specs to the agent local WorkQueue and update spec pickle file

**Please note:** This code will be used with polling cycle method, i.e. it will run as separate thread within WorkflowUpdater component. As such, it can have potentially impact on:
- upstream wmstats service, as we will perform concurrent requests to this service. The number of requests will correspond to total number of active workflows in a particular WMAgent. Since we have multiple agents this will spread across all of them.
   - for instance, we we have 100 active workflows at any given time these 100 calls will be spread across all agents
   - on wmstats side
- to update the spec we need to load pickle file, update its site lists, and write it back. This is potentially slow IO operation which will impact the component execution time:
   - if we'll have X active workflows in an agent it means and we'll spend O(sec) per pickle file load/save operation it means we'll spend X seconds in each agent during the update. The X number should be relatively small since not all active workflows may need site list change
   - it also means that our component can load X pickle files content into RAM to make an update
- if execution of update Site list function fails during polling cycle we expect that algorithm will pick it up in next cycle

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs

#### External dependencies / deployment changes

vkuznet committed:
New module to manage site list updates

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,109 @@
+        self.url = url
+        self.logger = logger
+
+    def getActiveRequests(self, wflowSpecs):
```
This should be using a SQL statement, similar to what you have implemented for pileup updates in WorkflowUpdater.

vkuznet replied:
If we'll put this class to be used in `WorkflowUpdaterPoller.py` (as I think it will be), then the `WorkflowUpdaterPoller.py` code does this SQL statement I didn't want to do it again as we already got wflowSpecs in here: https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/WorkflowUpdater/WorkflowUpdaterPoller.py#L286

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,109 @@
+                    ]
+        return requests, wmaSpecs
+
+    def getRequestSpecs(self, requests):
```
Even though we are going to need it to fetch the workflow spec itself, I think we will have to implement a call to the `wmstatserver` service and retrieve the site lists with a filter+mask call.

vkuznet replied:
Why? if we fetch the whole spec it contains site lists, why do we need to get it again from wmstatserver? I just don't understand the reason.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,109 @@
+                wmaSpec = wmaSpecs[key]
+                workloadHelper = WMWorkloadHelper()
+                workloadHelper.load(wmaSpec)
+                taskIterator = workloadHelper.taskIterator()
```
no need for iterating over the object ourselves. You can simply call the relevant WMWorkload object method, e.g.:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/WMWorkload.py#L676

vkuznet replied:
yes, we do need to iterate to check if sitelist needs to be changed, otherwise we only do assignment without knowing if it is required. But I don't mind to make the assignments. The reason I structured the code to identify which task must be updated and therefore when update happens we should push back the task to whatever place it should be.

todor-ivanov commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,31 @@ def cancelWorkflow(self, wf):
+                                 'reduce': False})
+        elementsToUpdate = [x['id'] for x in data.get('rows', [])]
+        if elementsToUpdate:
+            self.updateElements(*elementsToUpdate, SiteWhiteList=siteWhiteList, SiteBlackList=siteBlackList)
```
Just a note here, not a request for changing anything. When it comes to the global Workqueue we are about to attempt to create a method to update the workqueue elements per workflow : https://github.com/dmwm/WMCore/blob/1f78a64007e58f6c18aff2f0eeee0517d8d99927/src/python/WMCore/Services/WorkQueue/WorkQueue.py#L259-L291 
And  yet another method to update the full specs with all wmspec parameters at  a single push. For the later the following  methods was developed, but still not merged in the master branch:   https://github.com/dmwm/WMCore/blob/1f78a64007e58f6c18aff2f0eeee0517d8d99927/src/python/WMCore/WMSpec/WMWorkload.py#L78-L121


I know the scale here while working with the local queue is far from what we see in the global workqueue, but maybe once we converge on the other PR you may consider using the this same method here as well. If not anything else it at least

vkuznet replied:
ok, ,noted

todor-ivanov reviewed (approved):
It looks good to me.  I left just tow comments inline

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,167 @@
+        self.localWQ = WorkQueue(self.localCouchUrl,
+                                 self.config.WorkQueueManager.dbname)
+
+    def getActiveWorkflowss(self):
```
thank you for spotting it, fixed.

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,31 @@ def cancelWorkflow(self, wf):
+        data = self.db.loadView('WorkQueue', 'elementsDetailByWorkflowAndStatus',
+                                {'startkey': [wf], 'endkey': [wf, {}],
+                                 'reduce': False})
+        elementsToUpdate = [x['id'] for x in data.get('rows', [])]
```
I am thinking that we should filter these elements by Available and perhaps Acquired status. For now, can you filter them by the former only? Further thought is needed.

vkuznet replied:
I don't know how to apply filter, I followed logic of `updatePriority` function. Please point me to an example how filters are applied within CouchDB.

amaltaro replied:
To filter it in the query, I think we would have to pass in a tuple with the (wflow_name, status). However, I actually think we should use a cheaper view like:
https://cmsweb.cern.ch/couchdb/workqueue/_design/WorkQueue/_view/jobStatusByRequest?reduce=false

then iterate over the rows and compare the `Status` field with what we want.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+            # to one we received from upstream service (ReqMgr2)
+            wmaWhiteList = wHelper.getSiteWhiteList()
+            wmaBlackList = wHelper.getSiteBlackList()
+            if wmaWhiteList != siteWhiteList or wmaBlackList != siteBlackList:
```
We probably need to convert these to sets to have a correct list comparison (ordered list comparison).

vkuznet replied:
ok, converted to sets

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,31 @@ def cancelWorkflow(self, wf):
 
+    def updateSiteLists(self, wf, siteWhiteList, siteBlackList):
+        """
+        Update site lists of a workflow
```
Can you please improve this docstring? Please refer to https://github.com/dmwm/WMCore/blob/master/CONTRIBUTING.rst#project-docstrings-best-practices

In addition, I think we should make both sitewhitelist and siteblacklist optional, as a workflow could be updating only one of those lists.

vkuznet replied:
I'm not sure what to improve as docstring complaint with contributing requirements, i.e. it provides description of input parameters and what is returned by the function. If further improvements in docstring itself is required please be specific.

Meanwhile, I converted site/black lists to be optional parameters.

amaltaro replied:
Short description of the method can likely be better, e.g. "Update site list parameters in elements matching a given workflow and a list of element statuses.
In addition, parameters are supposed to mention the data type, as written in the document I linked above.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+                except Exception as ex:
+                    msg = f"Caught unexpected exception in SiteListUpdater. Details:\n{str(ex)}"
+                    logging.exception(msg)
+                    raise LocalWorkqueueUpdateException(msg) from None
```
If we raise this exception, we end up breaking the cycle. Instead, we should record the error but proceed with the other workflows in the pipeline.

vkuznet replied:
ok, switched to continue

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+
+                # update workload only if we updated local WorkQueue
+                # update site white/black lists together
+                wHelper.setWhitelist(siteWhiteList)
```
IMO, we should only call this if there was an actual change to these lists. The if statement above does not distinguish which of the 2 lists was changed.

vkuznet replied:
fixed

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+                except Exception as ex:
+                    msg = f"Caught unexpected exception in SiteListUpdater. Details:\n{str(ex)}"
+                    logging.exception(msg)
+                    raise SpecFileUpdateException(msg) from None
```
Same as the comment made for the other exception. We cannot break the whole cycle, instead, the component needs to keep processing further workflows and simply skip the one where we had problems updating the list.

vkuznet replied:
ok, switched to continue

amaltaro reviewed (changes requested):
Valentin, please find my comments along the code.

In addition to those, I would suggest to use `wmstatsserver` instead of ReqMgr2, such that we can drastically reduce the amount of HTTP calls made during each cycle.

An example url would be: https://cmsweb.cern.ch/wmstatsserver/data/filtered_requests?RequestStatus=running-open&mask=SiteWhitelist&mask=SiteBlacklist

and I would suggest to repeat it for `RequestStatus=acquired` as well.

For that, I think we can simply use this wrapper method:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/WMStatsServer/WMStatsServer.py#L92

instead of writing the pycurl calls from scratch.

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+from WMCore.WorkerThreads.BaseWorkerThread import BaseWorkerThread
+
+
+class SiteListUpdater(Harness):
```
ok, moved SiteListUpdatePoller into WorkflowUpdater.

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+        for wflowSpec in wflowSpecs:
+            name = wflowSpec['name']  # this is the name of workflow
+            pklFileName = wflowSpec['spec']  # the "spec" in WMBS table (wmbs_workflow.spec) is pkl file name
+            wmaSpecs[name] = pklFileName
```
I rather would not do it, it took me quite significant amount of time to figure out what kind/type of `wflowSpec['spec']` python object is and I made it explicit via proper naming convention to make other understand what it is. I suggest to keep it as I provided.

amaltaro replied:
Well, isn't that why you added the comment as well? Anyhow, discussing this thing is a waste of energy.

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+        self.logger = myThread.logger
+
+        # the reqmgr2Url should be points to ReqMgr2 data services, i.e. /reqmgr2 end-point
+        self.reqmgrUrl = getattr(config.SiteListUpdater, "reqmgr2Url")
```
ok, switching to wmstats then

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+    @timeFunction
+    def algorithm(self, parameters=None):
+        """
+        Executed in every polling cycle. The actual logic of the component is:
```
ok, done

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,194 @@
+
+                    # save back pickle file
+                    newPklFileName = pklFileName.split('.pkl')[0] + '_new.pkl'
+                    wHelper.save(newPklFileName)
```
the code is correct, the check is required as `newPklFileName` may not be written, e.g. disk is full, in fact we need even better check to see that the size of `newPklFileName` is comparable with  old one to make sure that file content is written.

amaltaro replied:
Valentin, we use this module for 10+ years. This code is very noisy and there is no need at all for all these concerns and checks. If the node runs out of filesystem storage, everything else would be broken, including SQL/NoSQL databases. Keep it simple please.

vkuznet commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,31 @@ def cancelWorkflow(self, wf):
+        # Update the spec, if it exists
+        if self.db.documentExists(wf):
+            wmspec = WMWorkloadHelper()
+            wmspec.load(self.hostWithAuth + "/%s/%s/spec" % (self.db.name, wf))
```
The WorkQueueManager in WMAgent points to local workqueue couchdb URL.

vkuznet committed:
Unit tests site list functionality

vkuznet commented:
test this please

amaltaro commented:
test this please

vkuznet commented:
test this please

amaltaro commented:
It looks like tests ran with your last trigger, as we can see here:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/15310/

but Jenkins failed to publish those results back to the PR. Maybe @d-ylee and/or @khurtado are doing something that might have cause this(?)

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , I give up with awaiting Jenkins post to this PR. I looked at https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/15310/ and style wise everything is fine. The reported unit tests failure do not seem to be relevant here either. Therefore I'm asking for your review again to speed up things.

amaltaro commented:
test this please

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
+"""
+File       : SiteListUpdater
```
Just to keep things more consistent across components, perhaps naming this as `SiteListPoller` would be better.

vkuznet replied:
done

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,164 @@
+        for state in states:
+            url = "{}/data/filtered_requests?RequestStatus={}&mask=SiteWhitelist&mask=SiteBlacklist".format(self.wmstatsUrl, state)
+            urls.append(url)
+        response = getdata(urls, ckey(), cert())
```
Please use the wrapper module for interacting with the WMStats Server. There is this API that reaches the endpoint you need:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/WMStatsServer/WMStatsServer.py#L92

vkuznet replied:
done, back to sequential API processing

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,164 @@
+            wmaWhiteList = wHelper.getSiteWhiteList()
+            wmaBlackList = wHelper.getSiteBlackList()
+            if set(wmaWhiteList) != set(siteWhiteList) or set(wmaBlackList) != set(siteBlackList):
+                self.logger.info(f"Updating {wflow}: siteWhiteList {wmaWhiteList} => {siteWhiteList} and siteBlackList {wmaBlackList} => {siteBlackList}")
```
I would suggest breaking this log record into multiple lines. Something like:
```
Updating {wflow}:
    SiteWhiteList {wmaWhiteList} => {siteWhiteList}
    SiteBlackList {wmaBlackList} => {siteBlackList}
```

vkuznet replied:
done

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListUpdater.py:
```diff
@@ -0,0 +1,164 @@
+                    # update local WorkQueue first
+                    self.localWQ.updateSiteLists(wflow, siteWhiteList, siteBlackList)
+                except Exception as ex:
+                    msg = f"Caught unexpected exception in SiteListUpdater. Details:\n{str(ex)}"
```
We are executing one method here, so I would suggest to make the error message less generic. E.g.: "Unexpected exception while updating elements in local workqueue....

vkuznet replied:
done

amaltaro reviewed (changes requested):
Valentin, changes are coming to a final shape. However, please look into my comments along the code.

amaltaro commented:
@khurtado did you hard-wire any tests to this specific PR? I am surprised to see that most of the other PRs are returning with jenkins report, but this one.

Looking into the log that was running this morning for this PR:
https://cmssdt.cern.ch/dmwm-jenkins/job/WMCore-PR-test-job-khurtado/621/console

I see specifically:
```
...
Setting status of 10bc178f7b2e453915fc1cf2e656d617d03482fe to PENDING with url https://cmssdt.cern.ch/dmwm-jenkins/job/WMCore-PR-test-job-khurtado/621/ and message: 'Build started for merge commit.'
...
Recording test results
ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?
Setting status of 10bc178f7b2e453915fc1cf2e656d617d03482fe to FAILURE with url https://cmssdt.cern.ch/dmwm-jenkins/job/WMCore-PR-test-job-khurtado/621/ and message: 'Build finished. '
```

These tests have been persistently failing for 5 days now. Can you please look into this?

amaltaro commented:
test this please

khurtado commented:
Regarding jenkins, the CC7 setup ran these tests:

https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCorePy3-PR-unittests/3503/
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-pylint3k/4447/
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-pylintpy3/3471/

However, the html report:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/15342/console

in the main build failed with:
```
Evaluating pylint report for file: pylintReport.json
File pylintReport.json not found.
Evaluating pylint report for file: pylintpy3Report.json
Traceback (most recent call last):
  File "scripts/PullRequestReport.py", line 266, in <module>
    "pylintpy3Report.json")
  File "scripts/PullRequestReport.py", line 71, in buildPylintReport
    float(report[filename]['base'].get('score', 0))):
KeyError: 'base'
```

vkuznet commented:
> Regarding jenkins, the CC7 setup ran these tests:
> 
> https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCorePy3-PR-unittests/3503/ https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-pylint3k/4447/ https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-pylintpy3/3471/
> 
> However, the html report: https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/15342/console
> 
> in the main build failed with:
> 
> ```
> Evaluating pylint report for file: pylintReport.json
> File pylintReport.json not found.
> Evaluating pylint report for file: pylintpy3Report.json
> Traceback (most recent call last):
>   File "scripts/PullRequestReport.py", line 266, in <module>
>     "pylintpy3Report.json")
>   File "scripts/PullRequestReport.py", line 71, in buildPylintReport
>     float(report[filename]['base'].get('score', 0))):
> KeyError: 'base'
> ```
> 
> This is the report I see in the CC9 environment: https://cmssdt.cern.ch/dmwm-jenkins/job/WMCore-PR-Report-khurtado/89/artifact/artifacts/PullRequestReport.html

@khurtado , so what we should do with it? Who suppose to fix it? Where?

I already submit my new changes 1/2h ago and again don't see any jenkins results here. How I can find out my jenkins build(s) each time I submit new changes?

vkuznet commented:
@amaltaro , I addressed your issues and mark them resolved but last three comments did not provide me any fields to reply and/or mark them as unresolved but I addressed all of them in a code. That said, jenkins results are still not posted to this PR. But I manually look them up [here](https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/15347/). Everything looks good from pylint and unit test points of view. Please proceed with review again.

khurtado commented:
@vkuznet  @amaltaro 
I created an issue describing the bug:

https://github.com/dmwm/WMCore/issues/12153

In short, the problem comes from the fact that a "score" is missing for the first file in the PR-pylintpy3 jenkins json report. This happens in CC7 (which code has not been manipulated), and the CC9 setup as well (since we just adapted the CC7 code to work on RHEL9).

https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-pylintpy3/3471/artifact/pylintpy3Report.json/*view*/

khurtado commented:
test this please

khurtado commented:
test this please

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+        """
+        Return list of requests specs for provided list of request names
+        :param requests: list of workflow requests names
+        :return wsList: list of workflow records obtained from wmstats server, each record has the following structure
```
In general we do not add the variable name that is returned, as also shown in the document I previously shared here.

vkuznet replied:
ok

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+        self.logger = myThread.logger
+
+        # the reqmgr2Url should be points to ReqMgr2 data services, i.e. /reqmgr2 end-point
+        self.wmstatsUrl = getattr(config.SiteListPoller, "wmstatsUrl")
```
In addition, this new variable and its value need to be defined in this file under the expected section:
https://github.com/dmwm/WMCore/blob/master/etc/WMAgentConfig.py

vkuznet replied:
ok, done

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+        {"RequestName": "bla", "SiteWhitelist":[], "SiteBlacklist": []}
+        """
+        # get list of workflows from wmstats
+        states = ['running-open', 'acquired']
```
I would suggest adding these states as a component variable in the init method. But we do not need to create a new component variable right now, we use the default value (these 2 above) but leave room for making it configurable in the component.

vkuznet replied:
ok

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+        wsList = []
+        for state in states:
+            inputConditions = {"RequestStatus": state}
+            resp = self.wmstatsSrv.getFilteredActiveData(inputConditions, outputMask)
```
Add a log info record before fetching this information, so we know we are running this I/O operation.

vkuznet replied:
ok, done

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+            wmaSpecs[name] = pklFileName
+        return wmaSpecs
+
+    def getRequestSpecs(self, requests):
```
The word "spec" seems to be slightly overloaded in this module.
Hence, I would suggest to rename this method with something more meaningful (e.g. getRequestSites)

vkuznet replied:
ok, changed.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+        Perform the following logic:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream ReqMgr2 server
+        - update site lists of all workflows
```
definitely not "all workflows"

vkuznet replied:
corrected.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+        wflows = wmaSpecs.keys()
+
+        # obtain workflow records from wmstats server
+        wsList = self.getRequestData(wflows)
```
Unless I am missing something, `getRequestData` is not defined.

vkuznet replied:
thanks, fixed the name of the function.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+
+        # iterate over workflow items and update local WorkQueue and pickle files if
+        # either site white or black lists are different
+        for rdict in wsList:
```
We should make sure that we iterate over the list of active workflows, which is supposed to be much smaller than the list of active workflows in WMStats. Otherwise the cost of loading each spec file will be unnecessarily high.

vkuznet replied:
ok, changes are made.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+                # update workload only if we updated local WorkQueue
+                # update site white/black lists together
+                if set(wmaWhiteList) != set(siteWhiteList):
+                    wHelper.setWhitelist(siteWhiteList)
```
I would add a log record here saying that we are updating the site whitelist for workflow xxx.
Similarly, for the site blacklist below

vkuznet replied:
ok

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+                    self.logger.info("Updating %s with new site lists for %s", self.localCouchUrl, wflow)
+                    wHelper.saveCouchUrl(self.localCouchUrl)
+
+                    # save back pickle file
```
I will insist on this. Please remove this code which is only adding noise IMO.
We load a spec with WMWorkloadHelper, we edit it, and we save it. No need to reinvent and/or overload with protections.

vkuznet replied:
ok.

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -203,7 +203,7 @@ def updateElements(self, *elementIds, **updatedParams):
         conflictIDs = self.db.updateBulkDocumentsWithConflictHandle(elementIds, eleParams, maxConflictLimit=20)
         if conflictIDs:
-            raise CouchConflictError("WQ update failed with conflict", data=updatedParams, result=conflictIDs)
+            raise CouchConflictError("WQ update failed with conflict", data=updatedParams, result=conflictIDs, status=400)
```
Why did you hardcode the 400 exit status here?

vkuznet replied:
Because `status` is mandatory parameter in `CouchCobnflictError`, see https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Database/CMSCouch.py#L1147 and because in this call it was not provided. The pylint correctly complaint about it.

amaltaro replied:
I don't think it is mandatory, as there is a default value in the parent class:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Database/CMSCouch.py#L1097

The 400 does seem to be the best choice as well, because this is not an user query error, it's just a concurrency error inherited to distributed systems. I would suggest not to change it, unless you are 100% sure that it is required (which I would be surprised to see why it didn't pop up before).

vkuznet replied:
According to class definition:
```
class CouchConflictError(CouchError):
    def __init__(self, reason, data, result, status):
        CouchError.__init__(self, reason, data, result, status)
        self.type = "CouchConflictError"
```
the constructor of `CouchConflictError` does not have status as optional parameter, it is mandatory. Therefore, either we need change ctor of this class to make `status=None` or `status=0` (last is more correct from data-type point of view) or as I did  (and pylint spot it correctly) we must provide `status` value during ctor call.

amaltaro replied:
You are right!
Then please change it to the correct status code, which according to CMSCouch is meant to be `409`

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,33 @@ def cancelWorkflow(self, wf):
+        # Update elements in Available status
+        data = self.db.loadView('WorkQueue', 'jobStatusByRequest',
+                                {'reduce': False})
+        states = ['Running']
```
This state should be Available

vkuznet replied:
ok, done

amaltaro commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -673,6 +673,15 @@ def removeTask(self, taskName):
 
+    def getSiteWhitelist(self):
+        """
+        Get the site white list from our data
```
"Get the site white list for the workflow"

vkuznet replied:
ok, done

amaltaro commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -689,6 +698,15 @@ def setSiteWhitelist(self, siteWhitelist):
 
+    def getSiteBlacklist(self):
+        """
+        Get the site black list from our data
```
"Get the site black list for the workflow"

vkuznet replied:
ok, done

amaltaro commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -689,6 +698,15 @@ def setSiteWhitelist(self, siteWhitelist):
+        Get the site black list from our data
+        :return: site black list
+        """
+        if getattr(self.data, 'SiteBlackList', None):
```
By the way, I think this code can be resumed to:
```
return getattr(self.data, "SiteBlackList", [])
```
similarly for the whitelist.

amaltaro reviewed (changes requested):
Valentin, sorry for the delay on getting back to you with this review.

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,156 @@
+        myThread = threading.currentThread()
+        self.logger = myThread.logger
+
+        # the reqmgr2Url should be points to ReqMgr2 data services, i.e. /reqmgr2 end-point
```
ok

vkuznet commented:
Alan, I addressed all your feedback items. Please review again.

amaltaro reviewed (changes requested):
Valentin, I really appreciate the fast turnaround you provide with these developments. Unfortunately, I turns to come with many typos and basic bugs that I would hope a reviewer would not spot, e.g. (at the very beginning of this review):
```
        self.wmstatsUrl = getattr(config.WorkflowUpdater, "wmstatsUrl")
        self.wmstatsSrv = WMStatsServer(config.wmstatsSvc_url, logger=self.logger)
```

For future cases, please either write the relevant unit tests, or cross-check your code before asking for a review. Once you are ready with that, please fire up another review request (and in addition to me, please include someone else).

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,33 @@ def cancelWorkflow(self, wf):
+        states = ['Available']
+        elementsToUpdate = [x['id'] for x in data.get('rows', []) if x['key'][-1] in states]
+        if elementsToUpdate:
+            self.updateElements(*elementsToUpdate, SiteWhiteList=siteWhiteList, SiteBlackList=siteBlackList)
```
Before the updateElements() call, it would be good to make a log record like "Updating %d elements in status %s for workflow %s"...

vkuznet replied:
ok, added

amaltaro commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -673,6 +673,21 @@ def removeTask(self, taskName):
+        Get the site white list for the workflow
+        :return: site white list
+        """
+        if getattr(self.data, 'SiteWhiteList', []):
```
Looking into the setter implementation, it looks like this if statement will never evaluate to true. Did you also observe it in the unit tests?

vkuznet replied:
removed

amaltaro reviewed (changes requested):
@vkuznet please find more comments along the code.

Please also update the initial PR description. I also have a question on:
> there is non-zero probability that local WorkQueue content for spec will deviate from pickle file if presented code will fail at run-time because local WorkQueue is separate service from actual pickle spec file on local file disk. At the moment there is no mechanism how to ensure that both operations will successfully persist with more complexity of the code, e.g. we need to record that local WorkQueue update was successful along with name of updated workflow(s) in order to update their pickle files later.

IIUC, a runtime error (or component shutdown) in the middle of the `algorithm` method could leave the workqueue couch spec up-to-date, while the filesystem spec would have the outdated lists. Is that a correct interpretation?

If so, the mechanism for resolving it is already implemented. As we are comparing the site lists from WMStats against the spec in the file system. So the agent would retry the same updates in the next cycle.

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+                self.logger.info("  siteBlackList %s => %s", wmaBlackList, siteBlackList)
+                try:
+                    # update local WorkQueue first
+                    self.localWQ.updateSiteLists(wflow, siteWhiteList, siteBlackList)
```
ok, done

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+        wflows = wmaSpecs.keys()
+
+        # obtain workflow records from wmstats server
+        wdict = self.getRequestSites(wflows)
```
ok, done

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+        """
+        Perform the following logic:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream ReqMgr2 server
```
ok, done

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+            wmaSpecs[name] = pklFileName
+        return wmaSpecs
+
+    def getRequestSites(self, requests):
```
ok, done

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+        self.logger = myThread.logger
+
+        # get wmstats parameters
+        self.wmstatsUrl = getattr(config.WorkflowUpdater, "wmstatsUrl")
```
ok, done

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+                try:
+                    # persist the spec in local CouchDB
+                    self.logger.info("Updating %s with new site lists for %s", self.localCouchUrl, wflow)
+                    wHelper.saveCouchUrl(self.localCouchUrl)
```
ok, done

vkuznet commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+                wflow = rdict.pop('RequestName')
+                # check that our workflow is in our requests list
+                if wflow in requests:
+                    wdict[wflow] = rdict
```
The data coming from this URL: https://cmsweb.cern.ch/wmstatsserver/data/filtered_requests?RequestStatus=running-open&mask=SiteWhitelist&mask=SiteBlacklist has this structure:
```
{"result": [
 {
  "SiteWhitelist": [
    "T1_US_FNAL"
  ],
  "SiteBlacklist": [],
  "RequestName": "cmsunified_ACDC1_Run2024G_Muon1_ZMu_PromptMUODPGNano_241105_094516_3263"
},{
  "SiteWhitelist": [
    "T1_US_FNAL"
  ],
  "SiteBlacklist": [],
  "RequestName": "cmsunified_ACDC1_Run2024G_Muon1_ZMu_PromptMUODPGNano_241105_094536_6961"
},....
```
And, the code has appropriate comment about it. The memory impact is very hard to measure as Python allocate memory based on amount of data it may need to hold. But I agree that dict assigned to dict may have negligible allocation. due to referencing.

vkuznet commented:
@amaltaro , this is kind ping that review is required.

amaltaro commented on src/python/WMComponent/WorkflowUpdater/SiteListPoller.py:
```diff
@@ -0,0 +1,152 @@
+
+                try:
+                    # persist the spec in local CouchDB
+                    self.logger.info("Updating %s with new site lists for %s", self.localCouchUrl, wflow)
```
Given that the pickle file is loaded from `pklFileName`, we might want to rephrase this record. Either just say we are updating it for workflow BLAH, or say that we are updating the pickle file at location XYZ.

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,34 @@ def cancelWorkflow(self, wf):
+        data = self.db.loadView('WorkQueue', 'jobStatusByRequest',
+                                {'reduce': False})
+        states = ['Available']
+        elementsToUpdate = [x['id'] for x in data.get('rows', []) if x['key'][-1] in states]
```
You need to filter elements that match the workflow name as well, otherwise in practice you would be updating **everything** under Available status in a given agent.

This url can give you an example output:
https://cmsweb-preprod.cern.ch/couchdb/workqueue/_design/WorkQueue/_view/jobStatusByRequest?reduce=false

vkuznet replied:
ok

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -237,6 +237,34 @@ def cancelWorkflow(self, wf):
+        states = ['Available']
+        elementsToUpdate = [x['id'] for x in data.get('rows', []) if x['key'][-1] in states]
+        if elementsToUpdate:
+            self.logger.info("Updating %d elements in status %s for workflow %s", wf)
```
You are missing arguments in here (spotted by Dario)

vkuznet replied:
ok

amaltaro commented on test/python/WMCore_t/WMSpec_t/WMWorkload_t.py:
```diff
@@ -336,6 +336,37 @@ def testDbsUrl(self):
 
+    def testGetSiteWhitelist(self):
+        """
+        Teats getSiteWhitelist and getSiteBlackList functionality of the task.
```
Typo

vkuznet replied:
ok

amaltaro reviewed (changes requested):
@vkuznet in general it is looking good, but there is one comment that must be resolved along the code (together with a few more interesting points).

vkuznet commented:
To the best of my knowledge this PR is ready for final review as its tests are passed (the failure in unit tests are not related to this PR) and all pylint errors are resolved along with reviewers comments.

amaltaro reviewed (commented):
@vkuznet these changes are now looking good to me. Before we can merge it, we still need to:
1) properly squash all these commits according to the guidelines
2) then I would like you to apply these changes to your test agent and have a workflow going through the system.

Right now, we cannot really test change of sites because the feature is not yet deployed in reqmgr2.

vkuznet commented:
@amaltaro , the changes are squashed
```
adf9ecdb6 (HEAD -> fix-issue-12039, origin/fix-issue-12039) Unit tests site list functionality
2bc5671bf New module to manage site list updates
```
and workflow has been submitted to vocms266, see `amaltaro_TaskChain_MC_vocms266_forceComplete_241120_204512_6352` in cmsweb-testbed.

vkuznet commented:
The [workflow](https://cmsweb-testbed.cern.ch/reqmgr2/fetch?rid=amaltaro_TaskChain_MC_vocms266_forceComplete_241120_204512_6352) has completed on patched vocms266 agent.

amaltaro reviewed (changes requested):
Valentin, I had a quick look at your agent logs, explicitly:
```
install/WorkflowUpdater/ComponentLog 
```

and I see the following crash starting up the new worker thread (SiteListPoller):
```
2024-11-25 20:40:38,811:140102459569984:CRITICAL:Harness:PostMortem: choked when initializing with error: 'ConfigSection' object has no attribute 'centralWMStatsURL'
  File "/usr/local/lib/python3.8/site-packages/WMCore/Agent/Harness.py", line 416, in startComponent
    self.prepareToStart()
  File "/usr/local/lib/python3.8/site-packages/WMCore/Agent/Harness.py", line 319, in prepareToStart
    self.preInitialization()
  File "/usr/local/lib/python3.8/site-packages/WMComponent/WorkflowUpdater/WorkflowUpdater.py", line 41, in preInitialization
    myThread.workerThreadManager.addWorker(SiteListPoller(self.config),
  File "/usr/local/lib/python3.8/site-packages/WMComponent/WorkflowUpdater/SiteListPoller.py", line 33, in __init__
    self.wmstatsUrl = getattr(config.WorkflowUpdater, "centralWMStatsURL")
```

It looks like the component configuration attribute is named differently. 

Please update this PR and retest your developments in your test agent node.

todor-ivanov reviewed (approved):
I have  nothing more to add on top of what has already been commented in this issue.

amaltaro reviewed (approved):
Thanks Valentin, these changes look good to me. Can you please squash the commits accordingly?

vkuznet commented:
@amaltaro , commits are squashed.
